[SPARK-20589] Allow limiting task concurrency per stage

dhruve opened the pull request:
## What changes were proposed in this pull request?
This change allows the user to specify the maximum no. of tasks running in a given job group. (Kindly see the jira comments section for more context on why this is implemented at a job group level rather than a stage level). This change is beneficial where the user wants to avoid having a DoS while trying to access an external service from multiple executors without having the need to repartition or coalesce existing RDDs.

This code change introduces a new user level configuration: `spark.job.[userJobGroup].maxConcurrentTasks` which is used to set the active no. of tasks executing at a given point in time.

The user can use the feature by setting the appropriate jobGroup and passing the conf:

```
conf.set("spark.job.group1.maxConcurrentTasks", "10")
...
sc.setJobGroup("group1", "", false)
sc.parallelize(1 to 100000, 10).map(x => x + 1).count
sc.clearJobGroup
```

#### changes proposed in this fix 
This change limits the no. of tasks (in turn also the no. of executors to be acquired) than can run simultaneously in a given job group and its subsequent job/s and stage/s if the appropriate job group and max concurrency configs are set.

## How was this patch tested?
Ran unit tests and multiple manual tests with various combinations of:
- single/multiple/no job groups
- executors with single/multi cores
- dynamic allocation on/off

dhruve committed:
[SPARK-20589] Allow limiting task concurrency per stage

dhruve commented:
Rebased this PR with current master and have squashed the earlier commits.

SparkQA commented:
**[Test build #81645 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81645/testReport)** for PR 19194 at commit [`4281151`](https://github.com/apache/spark/commit/4281151df9010b4e9fe91e588c07e872b8e0dd69).

SparkQA commented:
**[Test build #81645 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81645/testReport)** for PR 19194 at commit [`4281151`](https://github.com/apache/spark/commit/4281151df9010b4e9fe91e588c07e872b8e0dd69).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81645/
Test PASSed.

tgravescs commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -758,11 +812,58 @@ private[spark] class ExecutorAllocationManager(
+      // We can limit the no. of concurrent tasks by a job group. A job group can have multiple jobs
+      // with multiple stages. We need to get all the active stages belonging to a job group to
+      // calculate the total no. of pending + running tasks to decide the maximum no. of executors
+      // we need at that time to serve the outstanding tasks. This is capped by the minimum of no.
```
remove of in capped by the minimum of no.

dhruve replied:
okay.

tgravescs commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -758,11 +812,58 @@ private[spark] class ExecutorAllocationManager(
+        var runningTasks = 0
+        if (stageIdToTaskIndices.contains(stageId)) {
+          runningTasks =
+            stageIdToTaskIndices(stageId).size - stageIdToCompleteTaskCount.getOrElse(stageId, 0)
```
missing speculative tasks here in running.

dhruve replied:
Yes. Nice catch. We do need to account for all the tasks for a stage. And this should include speculative ones as well.

tgravescs commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -758,11 +812,58 @@ private[spark] class ExecutorAllocationManager(
+        totalPendingTasks(stageId) + runningTasks
+      }
+
+      def sumIncompleteTasksForStages: (Int, (Int, Int)) => Int = (a, b) => {
```
it would be nice to name variables something more useful then a and b.

tgravescs commented on core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala:
```diff
@@ -210,22 +216,282 @@ class ExecutorAllocationManagerSuite
+      .setMaster("myDummyLocalExternalClusterManager")
+      .setAppName("test-executor-allocation-manager")
+      .set("spark.dynamicAllocation.enabled", "true")
+      .set("spark.dynamicAllocation.testing", "true")
```
probably doesn't matter for the test but ideally for speculation on spark.speculation=true

tgravescs commented on core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala:
```diff
@@ -210,22 +216,282 @@ class ExecutorAllocationManagerSuite
+    assert(maxNumExecutorsNeeded(manager) === 2)
+
+    // submit a speculative task
+    sc.listenerBus.postToAll(SparkListenerSpeculativeTaskSubmitted(0))
```
this works because its pending, send the task started event for this and I think it will show the wrong value due to getIncompleteTasksForStage not handling speculative.

I think it would be good to have more tests due the task started to make sure the pending vs running logic works properly.

dhruve replied:
okay. I will add more unit tests to cover the scenario.

tgravescs commented on core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala:
```diff
@@ -210,22 +216,282 @@ class ExecutorAllocationManagerSuite
+    sc.listenerBus.postToAll(SparkListenerJobStart(2, 0, Seq(stage3), sc.getLocalProperties))
+    sc.listenerBus.postToAll(SparkListenerStageSubmitted(stage3))
+    assert(maxNumExecutorsNeeded(manager) === 33) // 5 + 11 + 17
+
```
send another speculative here, otherwise this test is same as the above one.
Have the speculative task finish from above.

dhruve replied:
okay.

tgravescs commented on core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala:
```diff
@@ -210,22 +216,282 @@ class ExecutorAllocationManagerSuite
+    // Mark job without job group as complete
+    sc.listenerBus.postToAll(SparkListenerStageCompleted(stage4))
+    sc.listenerBus.postToAll(SparkListenerJobEnd(4, 20, JobSucceeded))
+    assert(maxNumExecutorsNeeded(manager) === 17) // 350 - 333
```
update comment, you had 352 tasks with the speculative.

Can you perhaps finish a speculative here and have it go to 351 first.

tgravescs commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -1255,6 +1255,97 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+      "exec1" -> "host1",
+      "exec2" -> "host2"
+    ).flatMap { case (exec, host) =>
+      // offer each executor twice (simulating 2 cores per executor)
```
I think you are doing 3 core per executor here

dhruve replied:
yes. It should be 2.

dhruve committed:
Account for running speculative tasks

SparkQA commented:
**[Test build #81894 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81894/testReport)** for PR 19194 at commit [`04956cc`](https://github.com/apache/spark/commit/04956cc0ab84ca4291cbf119927f0ead74db4b31).

dhruve commented:
@tgraves I have addressed the comments and tried to cover the possible cases in the existing test for job groups and speculation. Kindly let me know if we need to add or address more use cases.

SparkQA commented:
**[Test build #81894 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81894/testReport)** for PR 19194 at commit [`04956cc`](https://github.com/apache/spark/commit/04956cc0ab84ca4291cbf119927f0ead74db4b31).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81894/
Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -758,11 +825,52 @@ private[spark] class ExecutorAllocationManager(
+
+      def doesSumOverflow(a: Int, b: Int): Boolean = b > (Int.MaxValue - a)
+
+      val stagesByJobGroup = stageIdToNumTasks.groupBy(x => jobIdToJobGroup(stageIdToJobId(x._1)))
```
you could just store `stageIdToJobGroupId`.  Simplifies this a bit, and then you dont' need to store `jobIdToJobGroup` at all, I think

dhruve replied:
I like the idea. I think this can be done. Will update the PR.

squito commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -1255,6 +1255,97 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+      "exec1" -> "host1",
+      "exec2" -> "host2"
+    ).flatMap { case (exec, host) =>
+      // offer each executor twice (simulating 3 cores per executor)
```
update comments
5 offers -> 6 offers
twice -> three times

dhruve replied:
okay. Won't be needed as will be removing the test case.

squito commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -1255,6 +1255,97 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+    assert(taskDescs.size === 1) // tsm accepts one as it can run one more task
+  }
+
+  test("do not limit max concurrent running tasks in a job group by default") {
```
I don't think this test really adds anything beyond other tests in this suite.

dhruve replied:
The previous test covers all the necessary checks introduced by the change. I added this to cover the default scenario when no job group is specified. Can do away with this.

squito commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -619,6 +625,47 @@ private[spark] class ExecutorAllocationManager(
+      }
+
+      jobIdToJobGroup(jobStart.jobId) = jobGroupId
+      if (!jobGroupToMaxConTasks.contains(jobGroupId)) {
```
this is probably a weird / unusual situation, but is this really the behavior you want if there are multiple jobs submitted for the same job group?  Wouldn't you just take the conf for the job group at the time each job was submitted?

Worst case with this approach: say you are *always* submitting multiple jobs for each job group; when one finishes, you immediately start another one, so that the new one partially overlaps the old one.  Then even if you change the conf, all jobs will keep using the old value forever.

dhruve replied:
From my understanding and in context of the current code, you would group jobs together when you want their concurrency to be limited. If you want different concurrency limits for different jobs, you would set them in a different jobgroup altogether. 

If there are multiple jobs in the same job group which run concurrently and one of them sets a value different, then which one wins for the existing jobs and the new job? If we want to have a different value for every job then the user would need a way to know and identify a spark job in his application code , probably by a job id. Only by means of identifying a job, would the user be able to set the config for that job. This cannot be known apriori and I don't know if there is an easy way that the user can know about the underlying spark job corresponding to the action. Hence we apply a setting at the jobgroup level which allows the user to allow him to control the concurrency without knowing the underlying job related details specific to spark in an easy manner.

Let me know if anything is unclear here or if you have more questions.

squito replied:
lemme put my concern another way:

why don't you remove the `if (!jobGroupToMaxConTasks.contains(jobGroupId))`, and just unconditionally always make the assignment `jobGroupToMaxConTasks(jobGroupId) = maxConTasks`?

that is simpler to reason about, and has all the properties we want.  I agree the scenario I'm describing is pretty weird, but the only difference I see between your version and this is in that scenario.  And its probably not the behavior we want in that scenario.

dhruve replied:
I understand your reasoning about setting the maxConTasks every time the job is set. However, I am not able to understand the scenario which you described. If a job completes and new one kicks off immediately, how does the new job partially overlap? Its only when all the stages & underlying tasks for the previous job have finished we would mark it as complete. So a new job won't overlap with a completed one. Am I missing something here?

squito replied:
You could submit jobs concurrently, from two different threads.  I didn't describe it very well -- say both threads are in the same job group.  Each thread can only do one job at a time, but maybe between the two threads, there is always some active job for the job group the entire time.

There are real use cases which are like this -- in "job-server" style deployments, there is a long-running spark context (most likely with cached RDDs), accepting requests from multiple users (perhaps sitting behind an http server).  Maybe some group of users are always put into one job group (eg., there is a low-priority group and a high priority group of users).  You might process more than one job for each group at a time, and there is a never ending stream of jobs.

(again, its not the most common scenario, but might as well have it behave correctly.)

dhruve replied:
Okay. We did consider the job-server style deployments, however I am not able to follow the example/scenario you mentioned.

So when you say that different set of users are assigned different groups, they have to be served by "different threads", because we can set the job group on a per thread basis only. 

While it is possible & valid for different threads to belong to the same job group, you cannot have multiple job groups for the same thread active simultaneously. 

So in this scenario where you would want to change the maxConcTasks for a given jobGroup in one thread, the changes would be visible in the other thread as well. Because both of them belong to the same job group. In this case, it seems that it wasn't configured correctly. I am still not able to follow how setting the value every single time will help here.

tgravescs replied:
If we are talking jobs within the same job groups, it seems like this would be very timing dependent as to what number you would get if you start allowing it to be changed real time.  Lets say you have 1 thread and set the job group.  Now if all the jobs within that group are launched serial then everything is easy, allowing it to be changed can make sense. But if from that thread you spawn other threads to launch jobs in parallel (which would still be in that same job group) and each of those is setting it differently, how do you know you will get the right number for each of those jobs?   the 2 threads could race to set the conf and if both set it right before launching you are going to get one of the settings for both launches whereas one might have expected a different setting.

@squito  does this cover the scenario you are referring to?  

while both of those cases might be rare, I would lean towards making sure its more predictable and only setting it once rather then having user get something they don't expect.  But either could probably be documented away if we see the serial type scenario being more beneficial.  

ideally it would be nice to set at the stage level but that is a lot more difficult.

squito replied:
Sorry I am doing a really bad job explaining my concerns.

Of course, if you have multiple threads racing to set the value for the same job group, its unpredictable which value you'll get.  Nothing we can do about that, its not the scenario I'm talking about.

Here's the step-by-step (hope I understand the proposed change correctly):

(1) T1: Set jobgroup = "foo".  Set maxConTasks=10.
(2) T1: Launch job 1.1  Uses maxConTasks=10
(3) T2: Set jobgroup = "foo".
(4) T2: Launch job 2.1. Uses maxConTasks=10
(5) T1:  Finish job 1.1.  Do *not* remove the entry for `jobGroupToMaxConTasks("foo")`, because there is still another job running for this job group in T2 (https://github.com/apache/spark/pull/19194/files#diff-b096353602813e47074ace09a3890d56R664)
(6) T1: Set maxConTasks=20
(7) T1: Launch job 1.2.  Uses maxConTasks=10, because `jobGroupToMaxConTasks.contains("foo")`, so we don't reset the value.
(8) T2: Finish job 2.1.  Again, do not remove `jobGroupToMaxConTasks("foo")`, as T1 is still running a job in this job group.
(9) Set maxConTasks=20
(10) T2: Run job 2.2.  Uses maxConTasks=10, because `jobGroupToMaxConTasks.contains("foo")`

As long as there a job running in T2 when T1 finishes its job (or vice versa), we never remove the prior value, and so never update it when starting a new job.  We could be stuck with maxConTasks=10 in both threads indefinitely, even though both threads have set maxConTasks=20.

If you remove that `if (!jobGroupToMaxConTasks.contains(jobGroupId))`, then when job 1.2 starts, you'd use the new value of maxConTasks=20.  The only weird thing is that job 2.1 suddenly switches mid-flight to using maxConTasks=20 as well.  But that seems more reasonable to me.

markhamstra replied:
Do we even need mutable state associated with a job group? Some things would be a lot simpler if maxConTasks could only be set when the job group is created; and if you need a different number of maxConTasks, then you have to use a different job group.

dhruve replied:
@squito Thanks for elaborating on this. I believe the last para of your explanation summarized your viewpoint. In a way that makes some sense because if you want to increase your max concurrent tasks, you know what you are doing, so if you see weird thing with other threads that you created, its fine. However, when you don't control the thread creation, I feel its best to set it just once to avoid the weirdness. Its much easier to use a different job group than explain one more weird behavior.

dhruve committed:
Map the jobgroupId to stageId instead of jobId

SparkQA commented:
**[Test build #82052 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82052/testReport)** for PR 19194 at commit [`dcfb861`](https://github.com/apache/spark/commit/dcfb86152c36032e9ec4f1545fb84c6f24287423).

SparkQA commented:
**[Test build #82052 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82052/testReport)** for PR 19194 at commit [`dcfb861`](https://github.com/apache/spark/commit/dcfb86152c36032e9ec4f1545fb84c6f24287423).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82052/
Test PASSed.

dhruve committed:
Resolve merge conflicts

SparkQA commented:
**[Test build #82090 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82090/testReport)** for PR 19194 at commit [`7d65a02`](https://github.com/apache/spark/commit/7d65a028ce9ef03cd8999b3f95c5e7916765e57e).

tgravescs commented:
Jenkins, test this please

SparkQA commented:
**[Test build #82092 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82092/testReport)** for PR 19194 at commit [`7d65a02`](https://github.com/apache/spark/commit/7d65a028ce9ef03cd8999b3f95c5e7916765e57e).

SparkQA commented:
**[Test build #82092 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82092/testReport)** for PR 19194 at commit [`7d65a02`](https://github.com/apache/spark/commit/7d65a028ce9ef03cd8999b3f95c5e7916765e57e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82092/
Test PASSed.

squito reviewed (changes requested):
I'm sorry, but I realized there is another problem here.  You can also have a stage shared by multiple jobs; those jobs may even have different job groups.

`DAGScheduler` already has some logic to try to deal with  this, see `activeJobForStage`.  I think the best thing to do is to take the properties of the active job when the stage is actually submitted, like the `DAGScheduler` does.

Doing that is tricky, though -- it doesn't look like the active job is exposed at all in `SparkListenerStageSubmitted`.  We could either (a) modify `SparkListenerStageSubmitted` to include the active job Id or (b) have `DAGScheduler` expose `activeJobForStage`.  I guess I prefer (a), in fact its probably useful for external consumers of the listener events.

squito commented:
one more thought -- have you considered adding stage properties?  I know that is more involved since its an API change, but I dunno how many times I've wanted something like that, eg. for labelling a stage in the UI, for partitioning hints, etc.  And if you had it, then it would remove all the complications in this of multiple jobs & threads etc., and allow the property to actually be set on just one stage, as opposed to the entire job, as was originally requested.

dhruve commented:
@squito Thanks for pointing that out. What you mentioned makes sense and I did dig some more on the `DAGScheduler` and `activeJobForStage` to gather more context. We could take into account the properties of the active job when the stage is submitted, however this behavior is indeterministic. 

Let's say if we have two jobs from two different job groups with different threshold of task concurrency, the one that's submitted first wins as the stage won't be recomputed for the second job. In this case, there is no control over which job can get submitted first before the second one (unless the user explicitly serializes them). The problem aggravates when the difference between the task concurrency threshold is large for the two jobs. In such a case, having a wrong value can completely take down your remote service.

For a deterministic behavior, I believe the best way to tackle this would be to handle in the stage properties as was the ask. However, since it involves an API change, I didn't go that route as the scope for that could be much broader. If we have more fundamental use cases which require adding something like this on the stage level, we should continue in that direction if the community is open and welcomes an API change.

squito commented:
btw I'm not opposed to this feature, just want to make sure that given (a) complexity and (b) the fact that it doesn't quite meet the original ask anyway (configuring at the stage & no need to break up pipelines), that the folks that wanted it are still for it.  I wonder if it should even be marked experimental in case we decide to drop it in favor of something stage level eventually.

squito commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -512,6 +535,9 @@ private[spark] class TaskSetManager(
       }
     } else {
+      if (runningTasks >= maxConcurrentTasks) {
+        logDebug("Already running max. no. of concurrent tasks.")
```
can you update this msg to include the jobGroup and full config name?

also I think it would be good to have this log at the info level one time (not on every `resourceOffer`, just the first time we hit the limit for this TaskSet).  I just see users hitting this inadvertently and so it would be good to make it clear.

dhruve replied:
I'll make the change for this and also update any comments to explain the behavior so far. Also I am not clear on the earlier reply as to what was the resolution for accounting the activeJobId. Do you still have any inputs ?

dhruve commented:
Configuring at the stage level seems to be the appropriate and more deterministic choice. If we agree on changing the API, we can start another effort looking in that direction. Till then we can mark this feature as experimental or have an undocumented config.

squito commented:
Sorry if I was unclear earlier on the issue w/ the active Job ID.  So, I agree, that if a user actually gets into this situation, where they've got two different jobs for the same stage, with different max concurrent tasks, its mostly a toss-up which one they'll get, as the users jobs are probably racing to get to that stage.  Still, I think its important that it pulls the max concurrent tasks from the active job, just so that users can understand what is going on and for consistency and debugability.  The TaskSetManager gets the property from the active job, which actually submitted the stage, so the ExecutorAllocationManager should do the same.

I think the best way to ensure that is to add activeJobId to SparkListenerStageSubmitted.  Then you'd go back to just keeping a jobIdToMaxConcurrentTasks map when handling onJobStart, and in onStageSubmitted, you'd then figure out the max number of tasks for that stage, given the job which actually submitted it.

@tgravescs what do you think?

tgravescs commented:
I agree with that its going to be a toss-up from the users perspective as to which setting will be applied.  But if the TaskSetManager is using the active job properties and that is what its basing the # of tasks to run on then we should match that same thing in the allocation manager so that we have the right number of executors to match what the task set manager is trying to run.

squito commented:
btw I'm going to have really intermittent access for a few weeks, so you don't need to wait for me to proceed with this @tgravescs

squito commented:
@dhruve is this still active?  Sorry I was out a while, catching up on everything.

dhruve commented:
Hi @squito. Sorry for the late response. I want to get back on this. As soon as I get a chance I will work on it and update the PR.

dhruve committed:
Resolve merge conflict

dhruve committed:
Update log message

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1062/
Test PASSed.

SparkQA commented:
**[Test build #93187 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93187/testReport)** for PR 19194 at commit [`b649906`](https://github.com/apache/spark/commit/b6499066612a025b88765e8d84b7ca79198c6a47).

dhruve commented:
@squito I have made the changes as requested. Can you have a look at this again. Thanks.

dhruve committed:
Resolve merge conflict

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1174/
Test PASSed.

SparkQA commented:
**[Test build #93350 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93350/testReport)** for PR 19194 at commit [`aac8a6a`](https://github.com/apache/spark/commit/aac8a6a619c8d60f66f9ddb072e0c4f9a7782621).

SparkQA commented:
**[Test build #93350 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93350/testReport)** for PR 19194 at commit [`aac8a6a`](https://github.com/apache/spark/commit/aac8a6a619c8d60f66f9ddb072e0c4f9a7782621).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93350/
Test FAILed.

tgravescs commented:
test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1188/
Test PASSed.

SparkQA commented:
**[Test build #93367 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93367/testReport)** for PR 19194 at commit [`aac8a6a`](https://github.com/apache/spark/commit/aac8a6a619c8d60f66f9ddb072e0c4f9a7782621).

SparkQA commented:
**[Test build #93367 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93367/testReport)** for PR 19194 at commit [`aac8a6a`](https://github.com/apache/spark/commit/aac8a6a619c8d60f66f9ddb072e0c4f9a7782621).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93367/
Test PASSed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6980/
Test PASSed.

squito commented:
oh my gosh, sorry @dhruve that I didn't see this for so long.  Is this still active?  If not, can you close?

squito commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -659,6 +677,35 @@ private[spark] class ExecutorAllocationManager(
+
+      if (jobGroupId == null || !conf.contains(s"spark.job.$jobGroupId.maxConcurrentTasks")) {
+        jobGroupId = DEFAULT_JOB_GROUP
+      }
```
this can be simplified to

```scala
val jobGroupId = if (jobStart.properties != null) {
        Option(jobStart.properties.getProperty(SparkContext.SPARK_JOB_GROUP_ID)).getOrElse(DEFAULT_JOB_GROUP)
      } else {
        DEFAULT_JOB_GROUP
      }
  val maxConTasks = conf.getOption(s"spark.job.$jobGroupId.maxConcurrentTasks").map(_.toInt).getOrElse(Int.MaxValue)
```

squito replied:
actually though I realize reading the rest of the comments, to make sure you get the properties from the job that actually ends up starting the stage, you shouldn't grab the properties here at all.  You should grab them from `stageSubmitted.properties`

squito commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -800,10 +874,47 @@ private[spark] class ExecutorAllocationManager(
+
+      def sumIncompleteTasksForJobGroup: (Int, (String, mutable.HashMap[Int, String])) => Int = {
+        (maxConTasks, x) => {
+          val totalIncompleteTasksForJobGroup = getIncompleteTasksForJobGroup(x._2.keySet.toSet)
```
I think this would be a little easier to follow as

```scala
def sumIncompleteTasksForJobGroup: (Int, (String, Set[Int])) => Int = {
  (sumSoFar, (jobGroup, stages) =>
    val totalIncompleteTasksForJobGroup = getIncompleteTasksForJobGroup(stages)
...
val stagesByJobGroup = stageIdToJobGroupId.groupBy(_._2).mapValues(_.keySet)
```

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15791/
Test PASSed.
